File: cpp/newton_raphson.cpp
// f(x)=x^3-10x^2+x+1

#include <iostream>
#include<iomanip>
#include<math.h>

using namespace std;

double f(double x) {
    return pow(x, 3) - 10 * pow(x, 2) + x + 1;
}

double df(double x) {
    return 3 * pow(x, 2) - 20 * x + 1;
}


int main()
{
    double x0 = 1.0;
    for(int i=0; i<100; i++){
        x0 = x0 - f(x0) / df(x0);
        cout << fixed << setprecision(10) << x0 << endl;

        if(abs(f(x0)) < 1e-10){
            break;
        }
    }

    return 0;
}

File: python/newton_raphson.py
# f(x)=x^3-10x^2+x+1

def f(x):
    return x**3 - 10*x**2 + x + 1


def df(x):
    return 3*x**2 - 20*x + 1


# x = 1
# for i in range(100):
#     x = x - f(x)/df(x)
#     print(x)
#     if abs(f(x)) < 1e-6:
#         break


x = 1
while abs(f(x)) > 1e-6:
    x = x - f(x)/df(x)
    print('x :', x, 'f(x) :', f(x))

